### PR TITLE
RPR2 optimizations for viewport

### PR DESCRIPTION
**Pull request by Vsevolod1983**
TCIKET: RPRMAYA-2341

PURPOSE:
Workflow is not optimized for usage of RPR2

EFFECT OF CHANGES:
Workflow is optimized for usage of RPR2
 - Render Callbacks
 - Multithreading processing
 - We update screen on every callback call we get from Core. So, from user perspective, if it drags mouse quite fast only center part of the screen will be updated. (It is slow if update entire screen).

### CODE REVIEW

**Review by bsavery — CHANGES_REQUESTED** [review]
Not a functional change, but please add a bit more detail to the effect description, do you want to wait for a review from @DMotorygin ?

**Comment by Vsevolod1983** [comment]
He had a vacation, but he is back fron today, so assigned

File: FireRender.Maya.Src/NorthStarRenderingHelper.cpp [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+{
+	m_pContext = pContext;
+	m_readBufferAndUpdateCallback = readBufferAndUpdateCallback;
+
+	m_pContext->SetRenderUpdateCallback(ContextRenderUpdateCallback, this);
+}
+
+void NorthStarRenderingHelper::OnContextRenderUpdateCallback(float progress)
+{
+	/*if (m_contextPtr->GetSamplesPerUpdate() == 1)
```
[the remark this answers is not in the document]
Comment (Vsevolod1983) [reply FireRender.Maya.Src/NorthStarRenderingHelper.cpp]: Done

File: FireRender.Maya.Src/NorthStarRenderingHelper.cpp [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+	m_DataReady = true;
+}
+
+void NorthStarRenderingHelper::UpdateThreadFunc()
+{
+	while (m_UpdateThreadRunning)
+	{
+		while (!m_DataReady && m_UpdateThreadRunning)
+		{
+			std::this_thread::sleep_for(1ms);
```
[the remark this answers is not in the document]
Comment (Vsevolod1983) [reply FireRender.Maya.Src/NorthStarRenderingHelper.cpp]: Done

File: FireRender.Maya.Src/NorthStarRenderingHelper.cpp [hunk 3]
```diff
@@ -66,30 +69,32 @@ void NorthStarRenderingHelper::SetData(FireRenderContext* pContext, std::functio
...
 	while (m_UpdateThreadRunning)
 	{
-		while (!m_DataReady && m_UpdateThreadRunning)
 		{
-			std::this_thread::sleep_for(1ms);
+			std::unique_lock<std::mutex> lck(m_DataReadyMutex);
+			m_DataReadyConditionalVariable.wait(lck, [this] { return !m_UpdateThreadRunning || m_DataReady; });
 		}
 
+		/*while (!m_DataReady && m_UpdateThreadRunning)
```
Comment (DMotorygin) [inline FireRender.Maya.Src/NorthStarRenderingHelper.cpp:93]: Please remove commented out code

File: FireRender.Maya.Src/Context/TahoeContext.cpp [hunk 4]
```diff
@@ -471,3 +471,11 @@ bool TahoeContext::IsGivenContextRPR2(FireRenderContext* pContext)
...
 	return pTahoeContext->m_PluginVersion == TahoePluginVersion::RPR2;
 }
+
+void TahoeContext::AbortRender()
+{
+	if (m_PluginVersion == TahoePluginVersion::RPR2)
+	{
+		GetScope().Context().AbortRender();
+	}
+}
```
Comment (DMotorygin) [inline FireRender.Maya.Src/Context/TahoeContext.cpp:481]: Please add new line at the end of file

File: FireRender.Maya.Src/NorthStarRenderingHelper.cpp (hunk 3 above)
Comment (Vsevolod1983) [reply FireRender.Maya.Src/NorthStarRenderingHelper.cpp]: Done

File: FireRender.Maya.Src/Context/TahoeContext.cpp (hunk 4 above)
Comment (Vsevolod1983) [reply FireRender.Maya.Src/Context/TahoeContext.cpp]: Done

**Review by DMotorygin — APPROVED** [review]

**Pull request merged**